### ISSUE TITLE
next-unread-mail jumps between first two mailboxes when imaps URL contains username/password

### ISSUE THREAD

**Issue by Bluewind** [neomutt/neomutt#1292]
## Expected Behaviour

When using the `next-unread-mailbox` command mutt/neomutt should go to the next mailbox with unread mails. For 5 mailboxes it should go through the list like this: 1,2,3,4,5,1,2,3,4,5,1,2,...

## Actual Behaviour

I have 5 mailboxes with unread mails, but `next-unread-mailbox` keeps jumping between INBOX and the first of the other mailboxes. It never gets to the remaining three. The order I see is: 1,2,1,2,1,2,1,2,...

## Steps to Reproduce

set folder=imaps://username@host/

Open an account with at least 3 mailboxes with unread mails. Try to jump to the `next-unread-mailbox` with the command.

## How often does this happen?

- Always

## When did it start to happen?

When I added a username (and password, but username is enough) to the imaps:// URL in the folder setting.

It does not happen when the username/password are set via imap_user/imap_pass and the folder URL  does not include them. 

## NeoMutt Version

```
NeoMutt 20180622
Copyright (C) 1996-2016 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 4.17.6-1-ARCH (x86_64)
ncurses: ncurses 6.1.20180127 (compiled with 6.1.20180127)
libidn: 1.34 (compiled with 1.34)
hcache backends: gdbm

Compiler:
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/8.1.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --enable-multilib --disable-werror --enable-checking=release --enable-default-pie --enable-default-ssp
Thread model: posix
gcc version 8.1.1 20180531 (GCC) 

Configure options: --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --gpgme --enable-lua --notmuch --gss --ssl --sasl --with-ui=ncurses --with-idn=/usr --gdbm

Compilation CFLAGS: -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include  -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens +getaddrinfo 
  -gnutls +gpgme +gss +hcache -homespool +idn -locales_hack +lua +meta 
  -mixmaster +nls +notmuch +openssl +pgp +sasl +smime +start_color 
  +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

```

## Extra Info

Arch Linux; happens in mutt and neomutt; Only tested with remote (IMAP) connection

**Comment by Bluewind** [comment]
Disclaimer: I just started using mutt so this comment might not be worded correctly since I know nothing about the inner workings of mutt.

Found an interesting workaround/detail. The issue only appears to happen for the mailboxes that are part of the current setting of folder. I have some macros to switch between different folder settings (all IMAP URLs) and if I use one of those, `next-unread-mailbox` works correctly for the mailboxes of the accounts that are not currently "active". Out of curiosity I tried the following config:

```
set folder=imaps://user:pass@host/
set folder=imaps://host/
```

First set the folder with username and password so that a connection can be opened. Then change the setting so it no longer includes the username/password. After that `next-unread-mailbox` works correctly even for the "active" account.

**Comment by Bluewind** [comment]
Why was this closed? I'm still experiencing this issue.

**Comment by flatcap** [comment]
Sorry.  Slightly over-zealous culling of old issues.

**Comment by Bluewind** [comment]
TL;DR: This appears to be fixed/no longer relevant in git master.

Note that I debugged this in the original mutt first due to #1435, which I believed to be another bug that I didn't want to tackle before my actual bug. The explanation below assumes that neomutt works the same as mutt internally, but I didn't actually attach a debugger to neomutt too since the bug is no longer reproducible in git master and reproducing in some older version seems pointless. The code looks quite similar though so I expect that it the same problem.

The bug was/is that `mutt_mailbox_next_buffer` searches for the currently open mailbox by going through the mailbox list and matching the URL/path against that of the open mailbox. However, the open mailbox path has the password in the URL, but the mailboxes it matches against have no password. Since it performs a simple string match, it obviously can't ever find a match and thus defaults to the start of the list. The exact line is https://github.com/neomutt/neomutt/blob/b37a784d4fc6efe74c121c01408e6fdb9cbb93fd/mailbox.c#L523-L525

Due to the change that prompted #1435, I've now removed my passwords from the URLs and started using account-hook instead. Obviously, the URLs can be properly matched by `mutt_mailbox_next_buffer` now since both URLs it compares are without the password.

PS: Neomutt's code is much, much easier to read than mutt's. Thanks for renaming `mutt_buffer_buffy` to `mutt_mailbox_next_buffer`! This actually explains what the method does. I wish I would have started gdb debugging with neomutt rather than mutt...

**Comment by flatcap** [comment]
> This appears to be fixed/no longer relevant in git master.

Great.  Thanks for letting us know.

> Neomutt's code is much, much easier to read than mutt's

Thank you :-)
I've spent a **LOT** of time tidying, documenting, renaming and refactoring.
